File: server/maths.go
package main

import (
	"math"
	"math/rand"
	"time"
)

type Point struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

type Vector struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

func MakePoint(x float64, y float64) *Point {
	return &Point{RoundToPlaces(x, 1), RoundToPlaces(y, 1)}
}

func Round(f float64) float64 {
	return math.Floor(f + 0.5)
}

func RoundToPlaces(f float64, places int) float64 {
	shift := math.Pow(10, float64(places))
	return Round(f*shift) / shift
}

func RoundPoint(p *Point) *Point {
	return &Point{RoundToPlaces(p.X, 1), RoundToPlaces(p.Y, 1)}
}

func RoundVector(v *Vector) *Vector {
	return &Vector{RoundToPlaces(v.X, 1), RoundToPlaces(v.Y, 1)}
}

// Converts an angle in degrees between 0 and 359.
func AngleToVector(angle float64) *Vector {
	// Convert to radians.
	r := angle * 0.01745
	return UnitVector(&Vector{X: math.Sin(r), Y: -math.Cos(r)})
}

func AngleAndSpeedToVector(angle float64, speed float64) *Vector {
	return MultiplyVector(AngleToVector(angle), speed)
}

func Magnitude(vector *Vector) float64 {
	return math.Sqrt(vector.X*vector.X + vector.Y*vector.Y)
}

func UnitVector(vector *Vector) *Vector {
	return &Vector{
		X: (vector.X / Magnitude(vector)),
		Y: (vector.Y / Magnitude(vector)),
	}
}

func MultiplyVector(vector *Vector, f float64) *Vector {
	return &Vector{
		X: vector.X * f,
		Y: vector.Y * f,
	}
}

func AddVectors(vector1 *Vector, vector2 *Vector) *Vector {
	return &Vector{
		X: vector1.X + vector2.X,
		Y: vector1.Y + vector2.Y,
	}
}

func AddVectorToPoint(vector *Vector, point *Point) *Point {
	return &Point{
		X: point.X + vector.X,
		Y: point.Y + vector.Y,
	}
}

func MakeTimestamp() uint64 {
	return uint64(time.Now().UnixNano() / int64(time.Millisecond))
}

func Random(min int, max int) int {
	d := max - min + 1
	return min + rand.Intn(d)
}

func RandomFloat(min float64, max float64) float64 {
	d := max - min + 1
	return min + rand.Float64()*d
}

func RandomAngle() float64 {
	return float64(Random(0, 359))
}

func DistanceBetweenPoints(p1 *Point, p2 *Point) float64 {
	dx := p1.X - p2.X
	dy := p1.Y - p2.Y
	return math.Sqrt(float64(dx*dx + dy*dy))
}

func CalculateCenter(points []*Point) *Point {
	var sx float64 = 0
	var sy float64 = 0
	for _, p := range points {
		sx += p.X
		sy += p.Y
	}
	return &Point{
		X: sx / float64(len(points)),
		Y: sy / float64(len(points)),
	}
}

func IsColliding(p1 *Point, r1 float64, p2 *Point, r2 float64) bool {
	return DistanceBetweenPoints(p1, p2) < (r1 + r2)
}

File: public/js/maths.js
// Lots of random math equations for figuring things out.


// Given a hex (#abcdef) brighten or darken by a percentage.
// Cribbed from http://stackoverflow.com/a/6444043/1063
function increaseBrightness(hex, percent) {
  // strip the leading # if it's there
  hex = hex.replace(/^\s*#|\s*$/g, '');

  // convert 3 char codes --> 6, e.g. `E0F` --> `EE00FF`
  if (hex.length == 3) {
    hex = hex.replace(/(.)/g, '$1$1');
  }

  var r = parseInt(hex.substr(0, 2), 16),
      g = parseInt(hex.substr(2, 2), 16),
      b = parseInt(hex.substr(4, 2), 16);

  return '#' +
    ((0|(1<<8) + r + (256 - r) * percent / 100).toString(16)).substr(1) +
    ((0|(1<<8) + g + (256 - g) * percent / 100).toString(16)).substr(1) +
    ((0|(1<<8) + b + (256 - b) * percent / 100).toString(16)).substr(1);
}

var utils = {
  getRandom: function(min, max) {
    return Math.random() * (max - min) + min;
  },
  angleToVector: function(angle) {
    // Convert to radians.
    var r = angle * 0.01745;
    return utils.unitVector({ x: Math.sin(r), y: -Math.cos(r) });
  },
  magnitude: function(vector) {
    return Math.sqrt(vector.x * vector.x + vector.y * vector.y);
  },
  unitVector: function(vector) {
    return {
      x: vector.x / utils.magnitude(vector),
      y: vector.y / utils.magnitude(vector)
    };
  },
  angleAndSpeedToVector: function(angle, speed) {
    return utils.multiplyVector(utils.angleToVector(angle), speed);
  },
  multiplyVector: function(vector, a) {
    return {
      x: vector.x * a,
      y: vector.y * a,
    };
  },
  simpleMovingAverage: function(period, toDiscard) {
    var nums = new Array(period);
    for (var i = 0; i < period; i++) {
      nums[i] = 0;
    }
    var length = 0;
    var idx = 0;
    var sum = 0;
    return function(num) {
      sum -= nums[idx]; // subtract last value
      nums[idx] = num; // record new value
      sum += num; // add new value
      if (length < period) { length++ }; // update length

      // advance pointer
      idx++;
      if (idx >= period) {
        idx = 0;
      }

      if (toDiscard > 0) {
        // discard the lowest and highest results in the set
        var sorted = nums.slice(0, length).sort();
        var n = Math.floor(toDiscard * length);
        var msum = 0;
        for (var i = n; i < (length - n); i++) {
          msum += sorted[i];
        }
        return msum / (length - n - n);
      } else {
        // return simple mean as result
        return sum / length;
      }
    }
  },
  roundToPlaces: function(f, places) {
    var shift = Math.pow(10, places);
    return Math.round(f*shift) / shift;
  },
}

var collision = {
  circles: function(x1, y1, r1, x2, y2, r2) {
    var dx = x2 - x1;
    var dy = y2 - y1;
    var distance = Math.sqrt(dx*dx + dy*dy);
    return (distance < r1 + r2);
  },

  squares: function(x2, y2, w2, h2, x2, y2, w2, h2) {
    var x_intersect = Math.abs(x1 - x2) * 2 < (w1 + w2);
    var y_intersect = Math.abs(y1 - y2) * 2 < (h1 + h2);
    return x_intersect && y_intersect;
  }
}
